src/Microsoft.PowerShell.Commands.Utility/commands/utility/WaitEventCommand.cs

```diff
@@ -84,13 +102,13 @@ protected override void ProcessRecord()
 
             while (!received)
             {
-                if (_timeoutInSeconds >= 0)
+                if (_timeoutTimespan.TotalMilliseconds >= 0)
                 {
-                    if ((DateTime.UtcNow - startTime).TotalSeconds > _timeoutInSeconds)
+                    if ((DateTime.UtcNow - startTime) > _timeoutTimespan)
                         break;
                 }
 
-                received = _eventArrived.WaitOne(200);
+                received = _eventArrived.WaitOne((int)(_timeoutTimespan.TotalMilliseconds / 100));
```

Comment:
The wait time should be the number of milliseconds until the expiry time occurs. So you wait exactly once, not a hundred times. So before entering the loop determined the point in time of when the expiry will occur, each time round the loop calculate the number of milliseconds between now and the expiry time. If the number of milliseconds is negative then you have expired.

If their is no expiry time set then WaitOne() should be used with no expiry time

Reply:
Unfortunately, it still has to check after each wait to see if the event has arrived.  Because this is called in a relatively tight loop, this will probably happen more than once.

After an even comes it, the cmdlet checks to see if it is the correct -SourceIdentifier.  If it was, then Wait-Event returns, otherwise, it checks again.

Reply:
That should not be right. It should be completely time independent except for the timeout. The AutoResetEvent _eventArrived  should be triggered in NotifyEvent which is called by ReceivedEvents_PSEventReceived which  is subscribed by Events.ReceivedEvents.PSEventReceived += ReceivedEvents_PSEventReceived;

So if there is no timeout then you should be able to use WaitOne() because the event will be triggered when there is a change.

Reply:
It should not be polling every some arbitrary time period, it should timeout when there is nothing to do, and it should wake up in the loop when an event is received to match the new event.

Reply:
@rhubarb-geek-nz noted, with two additional points:

1. The case I'm most concerned about is the rapid timeout or return of a single upcoming event (aka, providing a -Timespan).  I believe doing a slightly more efficient job of waiting forever does not help that scenario.
2. When the event is triggered, .WaitOne should return as soon as possible.  The timespan on .WaitOne is how long to wait before timing out.

If the PowerShell Team requests this change, we can address this.  Otherwise, I'd prefer to keep the code change small and the level of intrigue to a minimum, so as to maximize chances of a speedy integration.